src/workerd/api/pyodide/pyodide.h

```diff
@@ -65,12 +66,71 @@ class PyodideMetadataReader : public jsg::Object {
   }
 };
 
+// A loaded bundle of artifacts for a particular script id. It can also contain V8 version and
+// CPU architecture-specific artifacts. The logic for loading these is in getArtifacts.
+class ArtifactBundler : public jsg::Object {
+public:
+  ArtifactBundler(kj::Maybe<kj::Array<kj::byte>> existingSnapshot,
+      kj::Function<kj::Promise<bool>(kj::Array<kj::byte> snapshot)> uploadMemorySnapshotCb)
+      : existingSnapshot(kj::mv(existingSnapshot)),
+        uploadMemorySnapshotCb(kj::mv(uploadMemorySnapshotCb)),
+        hasUploaded(false) {};
+
+  ArtifactBundler()
+      : existingSnapshot(kj::none),
+        uploadMemorySnapshotCb(kj::none),
+        hasUploaded(false) {};
+
+  jsg::Promise<bool> uploadMemorySnapshot(jsg::Lock& js, kj::Array<kj::byte> snapshot) {
+    // Prevent multiple uploads.
+    JSG_REQUIRE(!hasUploaded, TypeError,
+        "This ArtifactBundler has already uploaded a memory snapshot");
```

Comment:
Actually, since this is returning a `jsg::Promise<bool>`, it would be better for these to return rejected promises rather than throwing synchronously.

e.g.

```
if (hasUploaded) {
  return js.rejectedPromise(js.typeError("This ArtifactBundle has already uploaded a memory snapshot");
}
```